proxy/agent.go

```diff
@@ -66,27 +67,58 @@ type Agent interface {
 	Run(stop <-chan struct{})
 }
 
+const (
+	defaultDelay = 1 * time.Hour
+)
+
 // NewAgent creates a new proxy agent for the proxy start-up and clean-up functions.
 func NewAgent(run func(interface{}, int) error, cleanup func(int),
-	maxRetries int, retryInterval time.Duration) Agent {
+	maxRetries int, initialInterval time.Duration) Agent {
 	return &agent{
-		run:           run,
-		cleanup:       cleanup,
-		epochs:        make(map[int]interface{}),
-		config:        make(chan interface{}),
-		status:        make(chan exitStatus),
-		maxRetries:    maxRetries,
-		retryInterval: retryInterval,
+		run:     run,
+		cleanup: cleanup,
+		epochs:  make(map[int]interface{}),
+		config:  make(chan interface{}),
+		status:  make(chan exitStatus),
+		retry: &retry{
+			delay:           defaultDelay,
+			defaultDelay:    defaultDelay,
+			budget:          maxRetries,
+			maxRetries:      maxRetries,
+			initialInterval: initialInterval,
+		},
 	}
 }
 
+type retry struct {
+	// delay between reconciliation adjusted by retries
+	delay time.Duration
+
+	// default delay is a long duration - reconciliation is a no-op in the regular case.
+	// For permanent errors, reconciliation is still attempted once after the default delay.
+	defaultDelay time.Duration
+
+	// number of times to attempts left to retry applying the latest desired configuration
+	budget int
+
+	// maximum number of retries
+	maxRetries int
+
+	// delay between the first restart, from then on it is multiplied by a factor of 2
+	// for each subsequent retry
+	initialInterval time.Duration
+}
+
 type agent struct {
 	// proxy start-up command
 	run func(interface{}, int) error
 
 	// proxy cleanup command
 	cleanup func(int)
 
+	// retry configuration
+	retry *retry
+
 	// desired configuration state
 	desired interface{}
```

Comment:
The name of "config" field in agent is bad for a channel. Can you rename it to something else, also rename desired --> desiredConf, current-->currentConf.